Update SingletonSymNode to allow more comparisons

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108316
* __->__ #108315

In this PR:
- {in,}equality between singleton and plain ints returns false instead of erroring
- Morally define the semantic of j0 > c to be as if j0 represented an array [s_0, s_1, ... s_n] and s_k > c for all k
- Just like for equality, we don't actually want to do the comparison one by one, instead j0 is constrained to some range [min, max]. By default this range is [2, int64_t::max] so that it acts like a size and passes 0/1 specialization checks.
- In the future, we can define some API to allow users to constrain the range of their singletons

soulitzer committed:
Update SingletonSymNode to allow more comparisons

[ghstack-poisoned]

soulitzer committed:
Update on "Update SingletonSymNode to allow more comparisons"


In this PR:
- {in,}equality between singleton and plain ints returns false instead of erroring
- Morally define the semantic of j0 > c to be as if j0 represented an array [s_0, s_1, ... s_n] and s_k > c for all k
- Just like for equality, we don't actually want to do the comparison one by one, instead j0 is constrained to some range [min, max]. By default this range is [2, int64_t::max] so that it acts like a size and passes 0/1 specialization checks.
- In the future, we can define some API to allow users to constrain the range of their singletons

TODO: Add some unit tests


[ghstack-poisoned]

soulitzer committed:
Update on "Update SingletonSymNode to allow more comparisons"


In this PR:
- {in,}equality between singleton and plain ints returns false instead of erroring
- Morally define the semantic of j0 > c to be as if j0 represented an array [s_0, s_1, ... s_n] and s_k > c for all k
- Just like for equality, we don't actually want to do the comparison one by one, instead j0 is constrained to some range [min, max]. By default this range is [2, int64_t::max] so that it acts like a size and passes 0/1 specialization checks.
- In the future, we can define some API to allow users to constrain the range of their singletons

TODO: Add some unit tests


[ghstack-poisoned]

soulitzer committed:
Update on "Update SingletonSymNode to allow more comparisons"


In this PR:
- {in,}equality between singleton and plain ints returns false instead of erroring
- Morally define the semantic of j0 > c to be as if j0 represented an array [s_0, s_1, ... s_n] and s_k > c for all k
- Just like for equality, we don't actually want to do the comparison one by one, instead j0 is constrained to some range [min, max]. By default this range is [2, int64_t::max] so that it acts like a size and passes 0/1 specialization checks.
- In the future, we can define some API to allow users to constrain the range of their singletons

TODO: Add some unit tests


[ghstack-poisoned]

soulitzer committed:
Update on "Update SingletonSymNode to allow more comparisons"


In this PR:
- {in,}equality between singleton and plain ints returns false instead of erroring
- Morally define the semantic of j0 > c to be as if j0 represented an array [s_0, s_1, ... s_n] and s_k > c for all k
- Just like for equality, we don't actually want to do the comparison one by one, instead j0 is constrained to some range [min, max]. By default this range is [2, int64_t::max] so that it acts like a size and passes 0/1 specialization checks.
- In the future, we can define some API to allow users to constrain the range of their singletons

TODO: Add some unit tests


[ghstack-poisoned]

soulitzer committed:
Update on "Update SingletonSymNode to allow more comparisons"


In this PR:
- {in,}equality between singleton and plain ints returns false instead of erroring
- Morally define the semantic of j0 > c to be as if j0 represented an array [s_0, s_1, ... s_n] and s_k > c for all k
- Just like for equality, we don't actually want to do the comparison one by one, instead j0 is constrained to some range [min, max]. By default this range is [2, int64_t::max] so that it acts like a size and passes 0/1 specialization checks.
- In the future, we can define some API to allow users to constrain the range of their singletons

TODO: Add some unit tests


[ghstack-poisoned]

soulitzer committed:
Update on "Update SingletonSymNode to allow more comparisons"


In this PR:
- {in,}equality between singleton and plain ints returns false instead of erroring
- Morally define the semantic of j0 > c to be as if j0 represented an array [s_0, s_1, ... s_n] and s_k > c for all k
- Just like for equality, we don't actually want to do the comparison one by one, instead j0 is constrained to some range [min, max]. By default this range is [2, int64_t::max] so that it acts like a size and passes 0/1 specialization checks.
- In the future, we can define some API to allow users to constrain the range of their singletons

TODO: Add some unit tests


[ghstack-poisoned]

soulitzer committed:
Update on "Update SingletonSymNode to allow more comparisons"


In this PR:
- {in,}equality between singleton and plain ints returns false instead of erroring
- Morally define the semantic of j0 > c to be as if j0 represented an array [s_0, s_1, ... s_n] and s_k > c for all k
- Just like for equality, we don't actually want to do the comparison one by one, instead j0 is constrained to some range [min, max]. By default this range is [2, int64_t::max] so that it acts like a size and passes 0/1 specialization checks.
- In the future, we can define some API to allow users to constrain the range of their singletons

TODO: Add some unit tests


[ghstack-poisoned]

ezyang commented on c10/core/ConstantSymNodeImpl.cpp:
```diff
@@ -1,23 +1,30 @@
+    TORCH_INTERNAL_ASSERT(other->singleton_int().has_value());         \
+    c10::raw::intrusive_ptr::incref(this);                             \
+    return other->OP(                                                  \
+        c10::intrusive_ptr<ConstantSymNodeImpl<T>>::reclaim(this));    \
```
This is old code, but it's a bit weird to me that you incref and then reclaim, I feel like there should be a more direct API for this.

soulitzer committed:
Update on "Update SingletonSymNode to allow more comparisons"


In this PR:
- {in,}equality between singleton and plain ints returns false instead of erroring
- Morally define the semantic of j0 > c to be as if j0 represented an array [s_0, s_1, ... s_n] and s_k > c for all k
- Just like for equality, we don't actually want to do the comparison one by one, instead j0 is constrained to some range [min, max]. By default this range is [2, int64_t::max] so that it acts like a size and passes 0/1 specialization checks.
- In the future, we can define some API to allow users to constrain the range of their singletons


[ghstack-poisoned]

soulitzer committed:
Update on "Update SingletonSymNode to allow more comparisons"


In this PR:
- {in,}equality between singleton and plain ints returns false instead of erroring
- Morally define the semantic of j0 > c to be as if j0 represented an array [s_0, s_1, ... s_n] and s_k > c for all k
- Just like for equality, we don't actually want to do the comparison one by one, instead j0 is constrained to some range [min, max]. By default this range is [2, int64_t::max] so that it acts like a size and passes 0/1 specialization checks.
- In the future, we can define some API to allow users to constrain the range of their singletons


[ghstack-poisoned]

soulitzer commented:
Updates:
- realized I was not handling the self->op(other) --> other->inverse_op(self) case properly, whoops! (now we have to support all 4 inequalities)
- incref + reclaim = reclaim_copy

soulitzer committed:
Update on "Update SingletonSymNode to allow more comparisons"


In this PR:
- {in,}equality between singleton and plain ints returns false instead of erroring
- Morally define the semantic of j0 > c to be as if j0 represented an array [s_0, s_1, ... s_n] and s_k > c for all k
- Just like for equality, we don't actually want to do the comparison one by one, instead j0 is constrained to some range [min, max]. By default this range is [2, int64_t::max] so that it acts like a size and passes 0/1 specialization checks.
- In the future, we can define some API to allow users to constrain the range of their singletons


[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge

soulitzer commented:
@pytorchbot merge
